pennylane/transforms/add_noise.py

```diff
@@ -198,10 +209,47 @@ def stop_at(obj):
                     curr_ops.extend(noise_ops)
         new_operations.extend(curr_ops)
 
-    new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)
-    post_processing_fn = qml.devices.preprocess.null_postprocessing
+    if not noise_model.meas:
+        new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)
+        return [new_tape], qml.devices.preprocess.null_postprocessing
+
+    meas_conds, meas_funcs = [], []
+    for condition, noise in noise_model.meas.items():
+        meas_conds.append(lru_cache(maxsize=512)(condition))
+        meas_funcs.append(qml.tape.make_qscript(noise))
+
+    new_tapes = []
+
+    split_operations, split_measurements = [], [[] for idx in tape.measurements]
+    for midx, measurement in enumerate(tape.measurements):
+        readout_operations = new_operations.copy()
+        for condition, noise in zip(meas_conds, meas_funcs):
+            if condition(measurement):
+                noise_ops = noise(measurement, **metadata).operations
+                readout_operations.extend(noise_ops)
+        if readout_operations not in split_operations:
+            split_operations.append(readout_operations)
+        split_measurements[split_operations.index(readout_operations)].append((midx, measurement))
+
+    split_measurements = split_measurements[: len(split_operations)]
+    split_meas_indexes = qml.math.argsort(
+        [m_ for ms in ([m[0] for m in meas] for meas in split_measurements) for m_ in ms]
+    )
+
+    new_tapes = [
+        type(tape)(operations, [meas[1] for meas in measurements], shots=tape.shots)
+        for operations, measurements in zip(split_operations, split_measurements)
+    ]
+
+    def post_processing_fn(results):
```

Comment:
Should this be private?

Reply:
Nope, as it won't be accessible to the users anyway.

Reply:
Then I suggest making it private to prevent confusion and accidental import in the future.

Reply:
Most of the transforms use this name and the [docs](https://github.com/PennyLaneAI/pennylane/blob/7efa9c4f9f263b49f4bb01195bc9e57f9d68ed00/pennylane/transforms/core/transform.py#L82) suggests using this as well. There won't be an accidental import associated with it as it is being defined within a function definition, so it is out-of-scope otherwise.

Reply:
👍